pkg/operator/ceph/cluster/mon/mon.go

```diff
@@ -365,42 +436,92 @@ func (c *Cluster) initClusterInfo(cephVersion cephver.CephVersion) error {
 	return nil
 }
 
-func (c *Cluster) initMonConfig(size int) (int, []*monConfig) {
-	mons := []*monConfig{}
+func (c *Cluster) initMonConfig(size int) (int, []*monConfig, error) {
 
 	// initialize the mon pod info for mons that have been previously created
+	mons := c.clusterInfoToMonConfig("")
+
+	// initialize mon info if we don't have enough mons (at first startup)
+	existingCount := len(c.ClusterInfo.Monitors)
+	for i := len(c.ClusterInfo.Monitors); i < size; i++ {
+		c.maxMonID++
+		zone, err := c.findAvailableZoneIfStretched(mons)
+		if err != nil {
+			return existingCount, mons, errors.Wrap(err, "stretch zone not available")
+		}
+		mons = append(mons, c.newMonConfig(c.maxMonID, zone))
+	}
+
+	return existingCount, mons, nil
+}
+
+func (c *Cluster) clusterInfoToMonConfig(excludedMon string) []*monConfig {
+	mons := []*monConfig{}
 	for _, monitor := range c.ClusterInfo.Monitors {
+		if monitor.Name == excludedMon {
+			// Skip a mon if it is being failed over
+			continue
+		}
+		var zone string
+		schedule := c.mapping.Schedule[monitor.Name]
+		if schedule != nil {
+			zone = schedule.Zone
+		}
 		mons = append(mons, &monConfig{
 			ResourceName: resourceName(monitor.Name),
 			DaemonName:   monitor.Name,
 			Port:         cephutil.GetPortFromEndpoint(monitor.Endpoint),
+			Zone:         zone,
 			DataPathMap: config.NewStatefulDaemonDataPathMap(
 				c.spec.DataDirHostPath, dataDirRelativeHostPath(monitor.Name), config.MonType, monitor.Name, c.Namespace),
 		})
 	}
-
-	// initialize mon info if we don't have enough mons (at first startup)
-	existingCount := len(c.ClusterInfo.Monitors)
-	for i := len(c.ClusterInfo.Monitors); i < size; i++ {
-		c.maxMonID++
-		mons = append(mons, c.newMonConfig(c.maxMonID))
-	}
-
-	return existingCount, mons
+	return mons
 }
 
-func (c *Cluster) newMonConfig(monID int) *monConfig {
+func (c *Cluster) newMonConfig(monID int, zone string) *monConfig {
 	daemonName := k8sutil.IndexToName(monID)
 
 	return &monConfig{
 		ResourceName: resourceName(daemonName),
 		DaemonName:   daemonName,
 		Port:         DefaultMsgr1Port,
+		Zone:         zone,
 		DataPathMap: config.NewStatefulDaemonDataPathMap(
 			c.spec.DataDirHostPath, dataDirRelativeHostPath(daemonName), config.MonType, daemonName, c.Namespace),
 	}
 }
 
+func (c *Cluster) findAvailableZoneIfStretched(mons []*monConfig) (string, error) {
+	if !c.spec.IsStretchCluster() {
```

Comment:
Is it really necessary? Will this be triggered if not stretch cluster?

Reply:
Missed?

Reply:
I decided to keep this one. It's called from a couple different places and it was simpler to check inside the method if it's a stretch cluster.